modules/programs/fish.nix

```diff
@@ -45,109 +21,163 @@ in
         description = ''
-          Shell Script code called during fish shell initialisation.
+          Shell script code called during fish shell initialisation.
         '';
+        type = types.lines;
       };
 
       loginShellInit = mkOption {
-        type = types.lines;
         default = "";
         description = ''
           Shell script code called during fish login shell initialisation.
         '';
+        type = types.lines;
       };
 
       interactiveShellInit = mkOption {
-        type = types.lines;
         default = "";
         description = ''
           Shell script code called during interactive fish shell initialisation.
         '';
+        type = types.lines;
       };
 
       promptInit = mkOption {
-        type = types.lines;
         default = "";
         description = ''
           Shell script code used to initialise fish prompt.
         '';
+        type = types.lines;
       };
 
     };
+
   };
 
   config = mkIf cfg.enable {
 
-    environment.systemPackages = [ pkgs.fish ];
-
-    environment.pathsToLink = [ "/share/fish" ];
-
-    environment.loginShell = "${shell}/bin/fish -l";
-    environment.variables.SHELL = "${shell}/bin/fish";
+    environment.etc."fish/foreign-env/shellInit".text = cfge.shellInit;
+    environment.etc."fish/foreign-env/loginShellInit".text = cfge.loginShellInit;
+    environment.etc."fish/foreign-env/interactiveShellInit".text = cfge.interactiveShellInit;
+    environment.etc."fish/foreign-env/extraInit".text= cfge.extraInit;
+
+    environment.etc."fish/nixos-env-preinit.fish".text = ''
+      # This happens before $__fish_datadir/config.fish sets fish_function_path, so it is currently
+      # unset. We set it and then completely erase it, leaving its configuration to $__fish_datadir/config.fish
+      set fish_function_path ${pkgs.fish-foreign-env}/share/fish-foreign-env/functions $__fish_datadir/functions
+     
+      # For "reasons not remembered," setEnvironment does not set PATH, so do it here
+      set PATH ${replaceStrings [":"] [" "] config.environment.systemPath} 
+
+      # source the NixOS environment config
+      fenv source ${config.system.build.setEnvironment}
+      fenv source /etc/fish/foreign-env/extraInit
+
+      # clear fish_function_path so that it will be correctly set when we return to $__fish_datadir/config.fish
+      set -e fish_function_path
+    '';
 
     environment.etc."fish/config.fish".text = ''
       # /etc/fish/config.fish: DO NOT EDIT -- this file has been generated automatically.
 
-      set fish_function_path $fish_function_path ${pkgs.fish-foreign-env}/share/fish-foreign-env/functions
-
-      set PATH ${replaceStrings [":"] [" "] config.environment.systemPath}
-
-      ${config.system.build.setEnvironment}
-
-      fenv source ${foreignEnv}
-      ${cfg.shellInit}
-
-      ${concatStringsSep "\n" fishVariables}
+      # if we haven't sourced the general config, do it
+      if not set -q __fish_nixos_general_config_sourced
```

Comment:
@LnL7 @cbarrett 

do we care that references to NixOS have now ended up in the module? Would it be worth adding a PR to Nixpkgs which changes these to generic names, perhaps substituting `nix` for `nixos` and makes a note that this path is depended-on in nix-darwin and potentially other declarative configuration/module systems for Nix and not just NixOS?

Is it worth it just for the name change? Is it worth it to leave a note to Nixpkgs/NixOS developers so they know to at least ping people should that change?

Reply:
Good point. I'm down with changing the names to `nix_darwin` instead of `nixos`. How do others feel?